crates/sui-indexer-alt/src/handlers/tx_calls_fun.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use diesel_async::RunQueryDsl;
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::transaction::TransactionDataAPI;
+
+use crate::{db, models::transactions::StoredTxCallsFun, schema::tx_calls_fun};
+
+use super::Handler;
+
+pub struct TxCallsFun;
+
+#[async_trait::async_trait]
+impl Handler for TxCallsFun {
+    const NAME: &'static str = "tx_calls_fun";
+
+    const BATCH_SIZE: usize = 100;
+    const CHUNK_SIZE: usize = 1000;
+    const MAX_PENDING_SIZE: usize = 10000;
+
+    type Value = StoredTxCallsFun;
+
+    fn process(checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
+        let CheckpointData {
+            transactions,
+            checkpoint_summary,
+            ..
+        } = checkpoint.as_ref();
+
+        let mut values = Vec::new();
+        let first_tx = checkpoint_summary.network_total_transactions as usize - transactions.len();
+
+        for (i, tx) in transactions.iter().enumerate() {
+            let tx_sequence_number = (first_tx + i) as i64;
+            let sender = tx.transaction.sender_address().to_vec();
+
+            let move_calls = tx.transaction.data().transaction_data().move_calls();
+            values.extend(
+                move_calls
+                    .iter()
```

Comment:
thoughts on doing something like
```
    let values: Vec<T> = transactions
        .iter()
        .enumerate()
        .flat_map(|(i, tx)| {
            let tx_sequence_number = (first_tx + i) as i64;
            T::extract_from_tx(tx, tx_sequence_number)
        })
        .collect();
``` and implementing extract_from_tx for StoredTxCallsFun and the other Stored structs

Reply:
I'm not sure this adds a whole whole lot, and it would require having to decide the type that we extract into (the collection of `T` that we return).

However, on this topic, the pattern of always having to calculate `first_tx` is already annoying me, and I wonder whether we can add a helper to `CheckpointData` to iterate over transactions with their sequence number.

Reply:
Per my comment on `tx_digests` though, I do tend to like the pattern of using `map`/`flat_map` where possible (but this is completely down to style, so no obligation to follow that here):

```rust
fn process(checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
    let CheckpointData {
        transactions,
        checkpoint_summary,
        ..
    } = checkpoint.as_ref();

    let first_tx = checkpoint_summary.network_total_transactions as usize - transactions.len();
    
    Ok(transactions
        .iter()
        .enumerate()
        .flat_map(|(i, tx)| {
            let calls = tx.transaction.data().transaction_data().move_calls();
            calls
                .iter()
                .map(|(package, module, func)| StoredTxCallsFun {
                    tx_sequence_number: (first_tx + i) as i64,
                    package: package.to_vec(),
                    module: module.to_string(),
                    func: func.to_string(),
                    sender: tx.transaction.sender_address().to_vec(),
                })
        })
}
```